### PR TITLE
Expose a public 'check' method

### PR DESCRIPTION
**Pull request by charliermarsh**
See: #271.

### CODE REVIEW

**Comment by charliermarsh** [comment]
@Seamooo - Is this roughly what's needed for the LSP? (Is there any way to get the LSP to pass us a path, even as a string, to include in the error messages?)

File: src/lib.rs [hunk 1]
```diff
@@ -15,3 +24,52 @@ pub mod printer;
 pub mod pyproject;
 mod python;
 pub mod settings;
+
+/// Run ruff over Python source code directly.
+pub fn check(contents: &str) -> Result<Vec<Message>> {
```
Comment (charliermarsh) [inline src/lib.rs:29]: With this, you can do:

```rust
use anyhow::Result;
use ruff;

fn main() -> Result<()> {
    println!("{:?}", ruff::check("def f(): x = 1")?);
    Ok(())
}
```

**Comment by charliermarsh** [comment]
@Seamooo - What are you planning to use to implement the LSP? https://github.com/ebkalderon/tower-lsp?

**Comment by Seamooo** [comment]
the lsp can definitely pass a path, the source requirement is due to the dynamic nature of the buffer updating. I've implemented a rough version of the lsp from scratch https://github.com/Seamooo/ruffd, getting this in would enable a proof of concept.

**Comment by charliermarsh** [comment]
That's awesome. And it's [here](https://github.com/Seamooo/ruffd/blob/main/ruffd-core/src/server_ops.rs#L62) that we can plug into ruff as a library?

**Comment by charliermarsh** [comment]
(Did you write that implementation from scratch? Is it based off anything else?)

**Comment by charliermarsh** [comment]
Two questions:

1. Should I change the method signature to include something like `path: &str` or even `path: Path`? Some of the internals like to have a `Path`, e.g., for formatting the `Message` body, but it could be worked around if it needs to be omitted.
2. How should we determine the `pyproject.toml` path here? Is there any way to base it on the user's current working directory? Or can the LSP path an absolute path to a file?

**Comment by charliermarsh** [comment]
(I went ahead and did (1); if we pass the absolute `Path`, it will in turn solve (2).)

**Comment by Seamooo** [comment]
> That's awesome. And it's [here](https://github.com/Seamooo/ruffd/blob/main/ruffd-core/src/server_ops.rs#L62) that we can plug into ruff as a library?

Exactly

**Comment by Seamooo** [comment]
> (Did you write that implementation from scratch? Is it based off anything else?)

it's implemented from scratch from the specification, with coc-nvim as the client to verify

**Comment by Seamooo** [comment]
There's potentially some ambiguity for solving the `pyproject.toml` path. The easiest would just be using the workspace root provided by the lsp client, but potentially there's cases where the workspace has nested configurations underneath it. For that case the implementation would be to recurse from the file path upwards until finding a `pyproject.toml` terminating at the workspace root. If the public method takes in `&Settings`, potentially the public api shouldn't have to account for it. This may also enable global configs outside of pyproject.toml that can create the settings object.

**Comment by charliermarsh** [comment]
> For that case the implementation would be to recurse from the file path upwards until finding a pyproject.toml terminating at the workspace root.

If we pass `check` a path to a file, then `find_project_root` will automatically do this.

**Comment by Seamooo** [comment]
image
looking good!

**Comment by charliermarsh** [comment]
Nice! That's awesome!

What does it take to enable others to use the LSP from VS Code and elsewhere?

**Comment by charliermarsh** [comment]
(E.g., do we need to publish this on some sort of extension marketplace?)

**Pull request merged**